### PR TITLE
[CUDA] drop CUDA 10 support, start supporting CUDA 12 (fixes #5789)

### LINKED ISSUE #5789: [RFC] Drop support for CUDA 10
## Summary

CUDA 10 is very old. And with CUDA 10 we found some compilation problems that do not occur in CUDA 11. See (https://github.com/microsoft/LightGBM/pull/5605#issuecomment-1471250105) for example. Dropping support for CUDA 10 may reduce maintenance and CI test cost. Just want to hear your ideas about this @guolinke @jameslamb @StrikerRUS @jmoralez.

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #5789.

* drops support for CUDA 10
* adds support for CUDA 12

### Notes for Reviewers

CUDA 12.2.2 is the latest release of the CUDA toolkit (https://developer.nvidia.com/cuda-toolkit-archive), but the latest official Docker image is 12.2.0 (https://catalog.ngc.nvidia.com/orgs/nvidia/containers/cuda), so I made the new job here `12.2.0`.

I think it's important the `method: source` CI job be tested against the latest version of CUDA, since as of now installing from source is the only way users can work with a CUDA-enabled version of the Python package.

### CODE REVIEW

File: .ci/check_python_dists.sh [hunk 1]
```diff
@@ -25,7 +25,7 @@ if [ $PY_MINOR_VER -gt 7 ]; then
         pydistcheck \
             --inspect \
             --ignore 'compiled-objects-have-debug-symbols,distro-too-large-compressed' \
-            --max-allowed-size-uncompressed '60M' \
+            --max-allowed-size-uncompressed '70M' \
```
Comment (jameslamb) [inline .ci/check_python_dists.sh:28]: With generating code for 2 more CUDA architectures, the linux CUDA wheels are now around 63 MB

```text
==================== running pydistcheck ====================

checking '/LightGBM/dist/lightgbm-4.1.0.99-py3-none-manylinux_2_27_x86_64.whl'
----- package inspection summary -----
file size
  * compressed size: 40.6M
  * uncompressed size: 63.3M
  * compression space saving: 35.9%
contents
  * directories: 0
  * files: 17 (1 compiled)
size by extension
  * .so - 64423.5K (99.3%)
  * .py - 404.7K (0.6%)
  * no-extension - 21.5K (0.0%)
  * .txt - 0.0K (0.0%)
  * .typed - 0.0K (0.0%)
```

Build link: https://github.com/microsoft/LightGBM/actions/runs/6186545556/job/16794443363?pr=6099

We could start dropping older architectures at some point if we think that's a problem, but I don't think that should even be considered until we start possibly distributing precompiled CUDA wheels via a repository like PyPI.

Until then, I think building for more architectures to support a wider range of GPUs is desirable, and a few more MB on disk for the wheel isn't a big concern.

**Comment by jameslamb** [comment]
I'd specifically like a review from @shiyu1994 before we merge this one, to be sure I'm not missing something. I don't know a lot about CUDA 😅

**Comment by shiyu1994** [comment]
@jameslamb Thanks for working on this! This is very helpful. We definitely need to support CUDA 12. I'll review this in the next few days.

**Comment by jameslamb** [comment]
Sure no problem, take your time!

**Review by shiyu1994 — APPROVED** [review]

**Comment by jameslamb** [comment]
thanks @shiyu1994 !

**Pull request merged**